### PR TITLE
Handle program state in the summary

### LINKED ISSUE #14981: Anomaly when stepping to end of section in CoqIDE
#### Description of the problem

The following gives an anomaly when loaded into CoqIDE and you jump to the end of the file.

```coq

Section Foo.

  Lemma bar : Bar.
  Proof.
    {
  Admitted.

End Foo.
```
Here is a backtrace
```
Anomaly "File "vernac/vernacextend.ml", line 77, characters 66-72: Assertion failed."
Please report at http://coq.inria.fr/bugs/.
Raised at Vernacextend.OutProg.cast in file "vernac/vernacextend.ml", line 77, characters 66-78
Called from Vernacinterp.interp_typed_vernac in file "vernac/vernacinterp.ml", line 24, characters 11-38
Called from Vernacinterp.interp_control.(fun) in file "vernac/vernacinterp.ml", line 199, characters 24-69
Called from Flags.with_modified_ref in file "lib/flags.ml" (inlined), line 17, characters 14-17
Called from Flags.with_option in file "lib/flags.ml" (inlined), line 26, characters 24-78
Called from Flags.silently in file "lib/flags.ml", line 64, characters 19-40
Re-raised at Exninfo.iraise in file "clib/exninfo.ml", line 81, characters 4-38
Called from Vernacinterp.interp_gen.(fun) in file "vernac/vernacinterp.ml", line 249, characters 18-43
Called from Vernacinterp.interp_gen.(fun) in file "vernac/vernacinterp.ml" (inlined), line 247, characters 21-276
Called from Vernacinterp.vernac_timeout in file "vernac/vernacinterp.ml" (inlined), line 62, characters 4-7
Called from Vernacinterp.interp_gen in file "vernac/vernacinterp.ml" (inlined), line 247, characters 6-279
Called from Vernacinterp.interp_gen in file "vernac/vernacinterp.ml" (inlined), line 247, characters 6-279
Called from Vernacinterp.interp in file "vernac/vernacinterp.ml" (inlined), line 261, characters 2-57
Called from Stm.stm_vernac_interp in file "stm/stm.ml", line 1034, characters 4-58
Re-raised at Exninfo.iraise in file "clib/exninfo.ml", line 81, characters 4-38
Called from Stm.Reach.known_state.reach.(fun) in file "stm/stm.ml", line 2154, characters 16-43
Called from Stm.State.define in file "stm/stm.ml", line 955, characters 6-10
Re-raised at Exninfo.iraise in file "clib/exninfo.ml", line 81, characters 4-38
Called from Stm.Reach.known_state.reach in file "stm/stm.ml", line 2296, characters 4-105
Called from Stm.observe in file "stm/stm.ml" (inlined), line 2397, characters 4-60
Called from Stm.finish in file "stm/stm.ml" (inlined), line 2408, characters 12-50
Called from Stm.process_transaction in file "stm/stm.ml", line 2666, characters 22-45
Re-raised at Exninfo.iraise in file "clib/exninfo.ml", line 81, characters 4-38
Called from Stm.observe in file "stm/stm.ml" (inlined), line 2404, characters 4-20
Called from Stm.finish in file "stm/stm.ml" (inlined), line 2408, characters 12-50
Called from Stm.process_transaction in file "stm/stm.ml", line 2666, characters 22-45
Re-raised at Exninfo.iraise in file "clib/exninfo.ml", line 81, characters 4-38
Called from Stm.handle_failure in file "stm/stm.ml", line 2509, characters 2-26
Called from Dune__exe__Idetop.add in file "ide/coqide/idetop.ml", line 85, characters 25-60
Called from Dune__exe__Idetop.eval_call.interruptible in file "ide/coqide/idetop.ml" (inlined), line 443, characters 12-15
Called from Dune__exe__Idetop.eval_call in file "ide/coqide/idetop.ml" (inlined), line 448, characters 20-37
Called from Xmlprotocol.abstract_eval_call in file "ide/coqide/protocol/xmlprotocol.ml" (inlined), line 690, characters 29-44
Called from Dune__exe__Idetop.eval_call in file "ide/coqide/idetop.ml" (inlined), line 471, characters 2-42
Called from Dune__exe__Idetop.loop.process_xml_msg in file "ide/coqide/idetop.ml" (inlined), line 529, characters 15-26
Called from Dune__exe__Idetop.eval_call in file "ide/coqide/idetop.ml" (inlined), line 471, characters 2-42
Called from Dune__exe__Idetop.loop.process_xml_msg in file "ide/coqide/idetop.ml" (inlined), line 529, characters 15-26
Called from Dune__exe__Idetop.eval_call in file "ide/coqide/idetop.ml" (inlined), line 448, characters 20-37
Called from Xmlprotocol.abstract_eval_call in file "ide/coqide/protocol/xmlprotocol.ml" (inlined), line 690, characters 29-44
Called from Dune__exe__Idetop.eval_call in file "ide/coqide/idetop.ml" (inlined), line 471, characters 2-42
Called from Dune__exe__Idetop.loop.process_xml_msg in file "ide/coqide/idetop.ml" (inlined), line 529, characters 15-26
Called from Dune__exe__Idetop.eval_call in file "ide/coqide/idetop.ml" (inlined), line 471, characters 2-42
Called from Dune__exe__Idetop.loop.process_xml_msg in file "ide/coqide/idetop.ml" (inlined), line 529, characters 15-26
Called from Dune__exe__Idetop.loop in file "ide/coqide/idetop.ml" (inlined), line 580, characters 4-40
Called from Coqtop.start_coq in file "toplevel/coqtop.ml" (inlined), line 78, characters 2-36
Called from Dune__exe__Idetop in file "ide/coqide/idetop.ml", line 631, characters 2-18
```

This anomaly was thrown in a Coq development I was working on, this is the reduced version.

#### Coq Version
8.15+alpha

### PR DESCRIPTION
**Pull request by SkySkimmer**
Instead of having some bugged adhoc handling in vernacstate.

Fix #14981

### CODE REVIEW

File: vernac/vernacstate.ml [hunk 1]
```diff
@@ -93,19 +93,16 @@ type t = {
   parsing : Parser.t;
   system  : System.t;              (* summary + libstack *)
   lemmas  : LemmaStack.t option;   (* proofs of lemmas currently opened *)
-  program : Declare.OblState.t NeList.t;    (* obligations table *)
   opaques : Opaques.Summary.t;     (* opaque proof terms *)
```
Comment (ejgallego) [inline vernac/vernacstate.ml:96]: I need to think about this, as in a sense these two are isomorphic, you are now just embedding the state into `System.t`.

**Comment by gares** [comment]
Can you explain why this fixes the bug?

File: vernac/vernacstate.ml [hunk 2]
```diff
@@ -93,19 +93,16 @@ type t = {
...
 }
 
 let s_cache = ref None
 let s_lemmas = ref None
-let s_program = ref (NeList.singleton Declare.OblState.empty)
 
 let invalidate_cache () =
   s_cache := None;
-  s_lemmas := None;
-  s_program := NeList.singleton Declare.OblState.empty
```
Comment (SkySkimmer) [inline vernac/vernacstate.ml:108]: This is the bug

**Comment by SkySkimmer** [comment]
ping @ejgallego I want to do 8.15.1 soon

**Comment by ejgallego** [comment]
I am still a bit too busy but will try to process this later today.

**Comment by ejgallego** [comment]
> I am still a bit too busy but will try to process this later today.

Given how much I still have to do not gonna happen today, but for sure I'll find time during the hackathon

**Comment by ejgallego** [comment]
I think I prefer https://github.com/coq/coq/pull/15729 , but let's see what @SkySkimmer thinks.

**Review by ejgallego — CHANGES_REQUESTED** [review]
Thinking about it more, this is a nack for me now; the idea we had with the proof state refactoring is that functions do need to declare how they manipulate proof state in their type. This goes against that spirit IIUC.

File: vernac/vernacinterp.mli [hunk 3]
```diff
@@ -27,3 +27,5 @@ val test_mode : bool ref
 (** Default proof mode set by `start_proof` *)
 val get_default_proof_mode : unit -> Pvernac.proof_mode
 val proof_mode_opt_name : string list
+
+val get_program_state : unit -> Declare.OblState.t NeList.t
```
Comment (ejgallego) [inline vernac/vernacinterp.mli:31]: This is in particular the bit that bothers me here.

Comment (ejgallego) [reply vernac/vernacinterp.mli:31]: If I understand you correctly. Here, now plugins and other users of the API can access this without having to declare it in the type.

Comment (SkySkimmer) [reply vernac/vernacinterp.mli:31]: We can expose just something to `check_solved_obligations` if you prefer.
Plugins could already link stm so there was no capability change imo.

File: vernac/vernacstate.mli [hunk 4]
```diff
@@ -50,8 +50,6 @@ type t =
   (** summary + libstack *)
   ; lemmas  : LemmaStack.t option
   (** proofs of lemmas currently opened *)
-  ; program : Declare.OblState.t NeList.t
-  (** program mode table. One per open module/section including the toplevel module. *)
```
Comment (ejgallego) [inline vernac/vernacstate.mli:54]: Following discussion on the alternative PR #15729 , this bit here is quite useful to document managers, for example to provide a richer goal and obligations list.

**Comment by SkySkimmer** [comment]
Let's keep this in mind if we get another bug with program state, otherwise no need to fix what's not broken.

**Pull request closed without merging**